### DISCUSSION TITLE
xstate/store internal improvements

### DISCUSSION THREAD

**Discussion by TkDodo** [statelyai/xstate#4998] [Ideas]
Hey @davidkpiano 👋 

I was looking at `@xstate/store` today and noticed some things that I didn't fully understand and I think could be improved to fix some edge cases and reduce bundle size. I'm happy to implement all or some of them, let me know what you think:

1. Why is `store` that can be passed to `useSelector` potentially `undefined`?

https://github.com/statelyai/xstate/blob/ccaeb445c9a5e5406c020b30856d51a2102d5fa9/packages/xstate-store/src/react.ts#L34

It's in the type signature, but there's no test and no documentation for passing `useSelector(undefined)`, and it also doesn't make much sense imo. Maybe there's something I'm missing but I think we could remove some of the `!store` checks if we removed `undefined` from the union.

2. There's no need to memoize `getSnapshot` for `useSyncExternalStore`

https://github.com/statelyai/xstate/blob/ccaeb445c9a5e5406c020b30856d51a2102d5fa9/packages/xstate-store/src/react.ts#L54

The only thing that needs to be memoized is the `subscribe` function, `getSnapshot` could just be an inline function.

3. Missing implementation of `getServerSnapshot`

Right now, `getServerSnapshot` is the same `boundGetSnapshot` implementation as `getSnapshot`:

https://github.com/statelyai/xstate/blob/ccaeb445c9a5e5406c020b30856d51a2102d5fa9/packages/xstate-store/src/react.ts#L59-L60

To avoid hydration mismatches during the first client render pass after SSR, `getServerSnapshot` should _always_ return the initial state of the store. I think we can easily get that by returning `store.getInitialSnapshot()`. instead of `getSnapshot`. I've done the same implementation to zustand: 

- https://github.com/pmndrs/zustand/pull/2277

4. get rid of `use-sync-external-store` shim

The shim is an external dependency that takes up a bit of bundle size. Since `@xstate/store` is only targeting React 18 and higher:

https://github.com/statelyai/xstate/blob/ccaeb445c9a5e5406c020b30856d51a2102d5fa9/packages/xstate-store/package.json#L57-L59

we could use the built in `useSyncExternalStore` to get rid of that dependency. I'm aware that there is no `with-selector` equivalent built-in, but I don't think we need it: We can just call the `selector` inside the `getSnapshot` function. This will certainly work for the default comparison mechanism, and if we want to support custom comparators, we can do what [zustand is doing with `useShallow`](https://github.com/pmndrs/zustand/blob/3467a7a634cca027144f5eb686f0108c01bf3e8d/src/react/shallow.ts#L10-L19) and simply store the previous result in a ref. zustand offloads this to user-land and also provides a shallow comparator, but I think we can also have this built-in and call it internally.

Let me know what you think of those changes.

**Comment by davidkpiano** [comment]
Hey, I agree with all of these points (cc. @Andarist for feedback).

For the 1st one, that's just copied over from the `useSelector(…)` implementation in `XState`. The reason the first argument may be `undefined` there is for better DX when selecting from a child actor which may be undefined, since conditional hooks aren't a thing in React.

But for `@xstate/store`, we may not run into those same issues. It's possible we'll introduce a way for stores to spawn other stores (which will be really interesting, like an actor-model based alternative to "slices"), and then the DX may be helpful:

```ts
const store = createStore(...);

function Comp() {
  const child = useSelector(store, s => s.context.someChild)
  // child may be undefined
  const childStuff = useSelector(child, s => s?.context)
}

**Reply by TkDodo** [reply]
opened a PR here:
- https://github.com/statelyai/xstate/pull/5000
